api/src/main/java/org/openmrs/module/reporting/report/ReportRequest.java

```diff
@@ -47,6 +47,7 @@ public class ReportRequest extends BaseOpenmrsObject {
 	private Date renderCompleteDatetime;
     private Integer minimumDaysToPreserve;
 	private String description;
+	private String clientTimezone;
```

Comment:
Added new field so that the requester is able to specify the timezone for the dates on the resulting report

Reply:
I'm a little wary about this, but I'm not sure I have a much better alternative to offer.  The two existing mechanisms that could best support this already on the ReportRequest are:

1. RenderingMode.  If we can adapt the RenderingMode argument or otherwise evolve RenderingMode to allow passing in a timezone, we could potentially use this.  But I don't think this will be that straightforward based on how RenderingMode was originally designed.

2. Parameters.  This might be a reasonable alternative.  Basically we could say that any parameter passed in which matches the name of an existing "[context value](https://github.com/openmrs/openmrs-module-reporting/blob/master/api/src/main/java/org/openmrs/module/reporting/evaluation/EvaluationContext.java#L102)" in the evaluation context would override that context value.  Then we could add a new default "context value" named "timezone" which would have a value of the server timezone.  And this could be overridden by adding a parameter with the same name to the report.  Then, we'd ensure this timezone context value from the evaluation context was used whenever dates are rendered.

If we don't think using Parameters in this way is right, another take on 2 would be to add a new column to ReportRequest, but instead of having it be something very specific named "clientTimezone", we'd make it generic and call it "contextValues" or something, and let it define a serialized map or properties of key/values.

What do you think about the second option above?

Reply:
I assume that adding a parameter (2) will be something like, when creating the request:
```
ReportRequest reportRequest = new ReportRequest();
reportRequest.setReportDefinition(new Mapped<ReportDefinition>(reportDefinition, Map.of("timezone", "Europe/Zurich")));
```
Then we would override the "timezone" context value [here](https://github.com/openmrs/openmrs-module-reporting/blob/c0264d4cc441685edf1b3332bfd464d69c8f2e42/api/src/main/java/org/openmrs/module/reporting/report/service/ReportServiceImpl.java#L452C3-L452C70
)

The default "timezone" would be set [here](https://github.com/openmrs/openmrs-module-reporting/blob/master/api/src/main/java/org/openmrs/module/reporting/evaluation/EvaluationContext.java#L102) by adding `addContextValue("timezone", ZoneId.systemDefault());`

This seems a good approach to me.  @mseaton do you think we can proceed with using the existing parameters as above or is it worth creating a new column "contextValues" on the ReportRequest?

Let me know if I've got something wrong or if you have any other input.

cc @ibacher

Reply:
Sort of.  Basically we already have code in a few places in the reporting module that allows context values and parameter values to be used when rendering a report.  For example, the `ExcelTemplateRenderer` runs all values through the `EvaluationUtil.evaluateExpression` method, which allows values to be converted using a defined syntax (note, we'd need to make a fix here to set the context values first, and then add parameter values, currently this is reversed incorrectly).  However, we aren't at all consistent about how these are used (and this goes along with the comment you made below) - we'd have to review all of the different renderers and other ways that report data is output (eg. from the reportingrest module, etc) in order to ensure this was handled consistently, which might be difficult.  So this idea probably needs to be reconsidered.